lib/operations/execute_operation.js

```diff
@@ -76,11 +83,59 @@ function executeOperation(topology, operation, callback) {
     const handler = makeExecuteCallback(resolve, reject);
 
     try {
-      return operation.execute(handler);
+      if (operation.hasAspect(Aspect.EXECUTE_WITH_SELECTION)) {
+        return executeWithServerSelection(topology, operation, handler);
+      } else {
+        return operation.execute(handler);
+      }
     } catch (e) {
       handler(e);
     }
   });
 }
 
+function executeWithServerSelection(topology, operation, callback) {
+  const readPreference = operation.readPreference || ReadPreference.primary;
+
+  function callbackWithRetry(err, result) {
+    if (err == null) {
+      return callback(null, result);
+    }
+
+    if (!isRetryableError(err)) {
+      return callback(err);
+    }
+
+    // select a new server, and attempt to retry the operation
+    topology.selectServer(readPreference, (err, server) => {
+      if (err) {
+        callback(err, null);
+        return;
+      }
+
```

Comment:
If I'm not mistaken, I think there needs to be a second `shouldRetryReads` check here in case we select a second server that cannot retry.

Reply:
That's not my understanding reading [spec](https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#selecting-the-initial-server). There are sections for "when to allow a retry", but no corresponding sections after the retry is happening. At the point the retry is happening, we should defer to driver to do proper command construction, and the server to error appropriately

Reply:
https://github.com/mongodb/specifications/blob/master/source/retryable-reads/retryable-reads.rst#3a-selecting-the-server-for-retry :

> If the driver cannot select a server for the retry attempt **or the newly selected server does not support retryable reads**, retrying is not possible and drivers MUST raise the original retryable error. In both cases, the caller is able to infer that an attempt was made.

Reply:
done!